editoast/src/views/projection.rs

```diff
@@ -301,6 +318,206 @@ pub fn train_projection_input_hash(
+    infra_id: i64,
+    trains_schedules: Vec<TrainSchedule>,
+    electrical_profile_set_id: Option<i64>,
+) -> Result<HashMap<i64, ProjectPathTrainResult>> {
+    let path_projection = PathProjection::new(&path.track_section_ranges);
+    let mut valkey_conn = valkey_client.get_connection().await?;
+
+    let infra = Infra::retrieve_or_fail(conn, infra_id, || TrainScheduleError::InfraNotFound {
+        infra_id,
+    })
+    .await?;
+
+    let (rolling_stocks, _): (Vec<_>, _) = RollingStockModel::retrieve_batch(
+        conn,
+        trains_schedules
+            .iter()
+            .map::<String, _>(|t| t.rolling_stock_name.clone()),
+    )
+    .await?;
+
+    let rolling_stock_length: HashMap<_, _> = rolling_stocks
+        .into_iter()
+        .map(|rs| (rs.name, rs.length))
+        .collect();
+
+    let simulations = train_simulation_batch(
+        conn,
+        valkey_client.clone(),
+        core_client.clone(),
+        &trains_schedules,
+        &infra,
+        electrical_profile_set_id,
+    )
+    .await?;
+
+    let (trains_hash_values, trains_details) =
+        extract_train_details(&infra, &trains_schedules, simulations, &path).await?;
+
+    // 1. Retrieve cached projection
+    let (miss_cache, mut hit_cache) =
+        retrieve_cached_projection(&mut valkey_conn, &trains_hash_values, &trains_details).await?;
+
+    // 2 Compute space time curves and signal updates for all miss cache
+    let (space_time_curves, signal_updates) = join!(
+        compute_batch_space_time_curves(&miss_cache, &path_projection),
+        compute_batch_signal_updates(
+            core_client.clone(),
+            &infra,
+            &path.track_section_ranges,
+            &path.routes,
+            &path.blocks,
+            &miss_cache
+        )
+    );
+    let signal_updates = signal_updates?;
+
+    // 3. Store the projection in the cache (using pipeline)
+    let trains_hash_values: HashMap<_, _> = trains_details
+        .iter()
+        .map(|t| t.train_id)
+        .zip(trains_hash_values)
+        .collect();
+    let mut new_items = vec![];
+    for train_id in miss_cache.iter().map(|t| t.train_id) {
+        let hash = &trains_hash_values[&train_id];
+        let cached_value: CachedProjectPathTrainResult = CachedProjectPathTrainResult {
+            space_time_curves: space_time_curves
+                .get(&train_id)
+                .expect("Space time curves not available for train")
+                .clone(),
+            signal_updates: signal_updates
+                .get(&train_id)
+                .expect("Signal update not available for train")
+                .clone(),
+        };
+        hit_cache.push((cached_value.clone(), train_id));
+        new_items.push((hash, cached_value));
+    }
+    valkey_conn.json_set_bulk(&new_items).await?;
+
+    let train_map: HashMap<i64, TrainSchedule> =
+        trains_schedules.into_iter().map(|ts| (ts.id, ts)).collect();
+
+    // 4.2 Build the projection response
```

Comment:
Step 4.2 but there is no step 4 or step 4.1. Maybe some of the comments need to be updated.